File: python/beambot.py
# -*- coding: utf-8 -*-
import socket, urllib2, os;
from time import gmtime, strftime;
'''
BeamBot remade.
The old BeamBot was a compilation of horrible scripts written when I really had no idea what I was doing.
By Dan Jones for #BeamNG
'''

class BeamBot:
    HostName = "irc.danieljon.es"; # the hostname/IP of the IRC server
    Port = 6667; # the port used
    NickName = "BeamBot"; # the bots nickname
    Channel = "#fun" # our default channel to join
    Connected = False; # we use this later when connecting
    ImportantPeople = ["daniel_j", "badSol", "akino_germany", "DarthCain"]; # people who can use special commands

class Logging:
    LogName = "BeamNG.txt" # the file we will log messages to
    CleanLine = "null"; # this will hold the lines we use for logging
    CurrentUser = "null"; # last user who sent a message
    FileLine = "null"; # this is our file line
Bot = BeamBot; # initialize our bot information class
Log = Logging; # initialize our logging class

def WelcomeMessage(): # this function will send new users a welcome message
    WelcomeFile = "welcome.txt"; # set our welcome file name
    try: # try to load our file
        WelcomeMessage = open(WelcomeFile).read(); # read our welcome file
        ToReturn = "PRIVMSG #fun :" + Log.CurrentUser + " " + WelcomeMessage; # send our welcome message
        #return "PRIVMSG " + Bot.Channel + " :" + WelcomeMessage;

        #print "sent: " + WelcomeMessage + " to : " + Log.CurrentUser

    except: # file not found
        print("Can't load the welcome file");



def CheckForCommands(DataLine): # check for commands
    CommandsFile = open("commands.txt", "a+"); # open our file that contains the commands
    for Log.FileLine in CommandsFile.readlines():
        if(Log.FileLine.split(":")[0].upper() == Log.CleanLine.split(" ", 1)[0].upper()): # if the command is found we will send the response to the channel
            CommandsFile.close();
            return "PRIVMSG " + Bot.Channel + " :" + Log.FileLine.split(":", 1)[1]; # return the message we will send to the channel
        Log.FileLine = CommandsFile.readline(); # read a new line into the buffer
    CommandsFile.close();
    return "";


def CheckForFunkyCommands(DataLine): # here we look for some funky commands 
    try: # make sure we have the required info
        PossiblyFunky = DataLine.split("!")[1].split(" ")[0];
        FunkyParameter = DataLine.split("!")[1].split(" ", 1)[1];
    except: # funky paramter missing
        return "PRIVMSG " + Bot.Channel + " :Command parameter missing. All '!' commands require atleast one parameter!";

    if(PossiblyFunky.upper() == "USER"): # user wants to find someones BeamNG account...
        ToReturn = "PRIVMSG " + Bot.Channel + " :" + "http://www.beamng.com/member.php?username=" + FunkyParameter.replace(" ", "%20"); # set our return variable
        Log.FileLine = ToReturn;
        return ToReturn;

    if(PossiblyFunky.upper() == "ADDCOMMAND"): # add a command
        if(Log.CurrentUser in BeamBot.ImportantPeople): # is the user important?
            if(FunkyParameter[0] != "?"): # a ? is needed
                FunkyParameter = "?" + FunkyParameter; # append a ?
            if(FunkyParameter.find(":") == -1): # no text, the user is an idiot
                ToReturn = "PRIVMSG " + Bot.Channel + " :" + "You aren't very good at this. Example: !addcommand ?command:text to add"; # return message
            else:
                File = open('commands.txt','a') # open the file
                File.write('\n' + FunkyParameter) # write the new command
                File.close() # close file
                ToReturn = "PRIVMSG " + Bot.Channel + " :" + "Command " + str(FunkyParameter.split(":", 1)) + " added to the list"; # return message

        else:
            ToReturn = "PRIVMSG " + Bot.Channel + " :" + "You can't do this, I have no idea who you are."; # return message
        Log.FileLine = ToReturn;
        return ToReturn;

    if(PossiblyFunky.upper() == "REMOVECOMMAND"): # uremove a command
        Check = 0; # check if a command was deleted
        if(Log.CurrentUser in BeamBot.ImportantPeople): # is the user important?
            log = open("new.txt", "w"); # new file
            for line in open("commands.txt"): # all commands
                print line.split(":", 1)[0];
                print FunkyParameter;
                if not FunkyParameter.upper() in line.split(":", 1): # find the one selected
                    log.write(line) # write to our new file
                    log.close; # close file 
                else:
                    Check+= 1;
            if(Check < 1): # no command deleted
                ToReturn = "PRIVMSG " + Bot.Channel + " :" + "Command not found in the list"; # return message
            else:
                os.remove("commands.txt"); # remove our old commands file
                os.rename("new.txt", "commands.txt"); # rename our new file
                ToReturn = "PRIVMSG " + Bot.Channel + " :" + "Command (" + FunkyParameter + ") removed."; # return message
        else:
            ToReturn = "PRIVMSG " + Bot.Channel + " :" + "You can't do this, I have no idea who you are."; # return message
        Log.FileLine = ToReturn;
        return ToReturn;

def ParseRawDataAndLog(DataLine): # this function will return a clean string to log with etc
    try:
        Log.CurrentUser = DataLine.split("!")[0][1:]; # last user to send a message
        Log.CleanLine = DataLine.split("@")[1].split(":", 1)[1]; # the message sent
        if(DataLine.split("@")[1].split(":")[0][-6:].find("JOIN") != -1): # user joined the channel
            print(Log.CurrentUser + " Joined the channel\n");
            LoggingFile = open(Log.LogName, 'a+'); # open our log for writing, ONLY ONCE!
            LoggingFile.write("(" + Log.CurrentUser + " joined the channel)\n"); # write to our logging file
            LoggingFile.close(); # close our file after writing the line
            WelcomeMessage(); # send the user the welcome message


        if(DataLine.split("@")[1].split(":", 1)[0].find("KICK") != -1): # user was kicked
            print(DataLine.split("#")[1].split(" ")[1].split(" ")[0] + " was kicked from the channel. Reason: " + Log.CleanLine + "\n");
            LoggingFile = open(Log.LogName, 'a+'); # open our log for writing, ONLY ONCE!
            LoggingFile.write("(" + DataLine.split("#")[1].split(" ")[1].split(" ")[0] + " was kicked from the channel. Reason: " + Log.CleanLine + ")\n"); # write to our logging file
            LoggingFile.close(); # close our file after writing the line

        if((DataLine.split("@")[1].split(":", 1)[0].find("QUIT") != -1) or (DataLine.split("@")[1].split(":", 1)[0].find("PART") != -1) or (DataLine.split("@")[1].split(":", 1)[0].find("TIMEOUT") != -1)): # user quit
            print(Log.CurrentUser + " Left the channel.\n");
            LoggingFile = open(Log.LogName, 'a+'); # open our log for writing, ONLY ONCE!
            LoggingFile.write("(" + Log.CurrentUser + " Left the channel.)\n"); # write to our logging file
            LoggingFile.close(); # close our file after writing the line

        # Don't juge this line, he's a very special retard
        if(Log.CurrentUser != Bot.NickName and DataLine.split("@")[1].split(":", 1)[0][-6:].find("JOIN") == -1 and DataLine.split("@")[1].split(":")[0].find("KICK") == -1): # we really want to ignore the bot...
            print("<" + Log.CurrentUser + "> " + " " + Log.CleanLine);
            LoggingFile = open(Log.LogName, 'a+'); # open our log for writing, ONLY ONCE!
            LoggingFile.write(strftime("%H:%M:%S", gmtime()) + " <" + Log.CurrentUser + "> " + " " + Log.CleanLine + "\n"); # write to our logging file
            LoggingFile.close(); # close our file after writing the line


    except: # line confused as a channel message, simply ignore
        print("Line confused as channel message, ignoring...\n");

print("Connecting as: " + Bot.NickName + " to: " + Bot.HostName + " : " + str(Bot.Port));

'''
at this point we are ready to create and connect to our socket
we will then run a loop that will recieve information, handle PING/PONG, log messages, parse commands from file etc
'''
try:
    Socket = socket.socket(); # setup our socket for connections
    Socket.connect((Bot.HostName, Bot.Port)); # connect to our IRC server
    Data = Socket.makefile('r', 4096); # our data stream
    MessageToSend = "";
except:
    print("Connection error.\nAre you sure " + Bot.HostName + " has an IRC daemon running on port " + str(Bot.Port) + "?");
    raise SystemExit(0); # stop the bot at this point
# We are connected to the server, we must now send IMPORTANT user information to be accepted
Socket.send("NICK " + Bot.NickName + "\r\n"); # auth with our nickname
Socket.send("USER " + Bot.NickName + " " + Bot.NickName + " " + Bot.NickName + " :" + Bot.NickName + "\r\n"); # auth our real name etc
while True: # this loop will run forever
    DataLine = Data.readline().strip(); # this will hold our raw line of data
    if(DataLine.find("PING")) != -1: # detected ping
        Socket.send("PONG :" + DataLine.split(":")[1]); # send pong
        if(not Bot.Connected): # if we are not connected, we need to auth
            print("Authenticating and joining " + Bot.Channel + "..");
            Socket.send("PRIVMSG R : Login <>\r\n");
            Socket.send("MODE " + Bot.NickName + " +x\r\n"); # set mode +x
            Socket.send("JOIN " + Bot.Channel + "\r\n"); # join our channel
            Bot.Connected = True; #set the bot as connected
    if(len(DataLine) > 1 and Bot.Connected and DataLine.find("!") != -1):
        ParseRawDataAndLog(DataLine); # this will return a clean string that we can use for commans and logging
        if(Log.CleanLine[:1].find("?") != -1): # if the line is a POSSIBLE command, look for it
            MessageToSend = CheckForCommands(DataLine); # this will call our function that will check a list of commands for one matching what a user has typed
           
        if(Log.CleanLine[:1].find("!") != -1 and len(Log.CleanLine) > 4): # this line could POSSIBLY be a funky command, lets get FUNKAH
            MessageToSend = CheckForFunkyCommands(Log.CleanLine); # lets look for some funky commands
            
        if(MessageToSend != ""): # make sure we have a message to send
            try:
                Socket.sendall("\n" + MessageToSend + "\n"); # send the message to the channel
                LoggingFile = open(Log.LogName, 'a+'); # open our log for writing, ONLY ONCE!
                LoggingFile.write(strftime("%H:%M:%S", gmtime()) + "<" + Bot.NickName + "> " + " " + Log.FileLine.split(":", 1)[1]); # write to our logging file
                LoggingFile.close(); # close our file after writing the line
            except:
                print("Unknown error, ignoring everything.");
            MessageToSend = "";

File: c/beambot.c
/*
   Big thanks to http://www.patlathem.com/hello-irc-world/ for basically teaching me what to do. 
   BeamBot this time made in C!
   I think I am on the fourth complete BeamBot rewrite in as many years..
   But this time I know what I am doing (kind of, I did say this last time)
   ~ daniel_j
 */
#include <stdio.h> /* standard io */
#include <stdlib.h> /* standard library */
#include <string.h> /* string */
#include <sys/socket.h> /* sockets */
#include <netdb.h> /* network host */
#define MAXLINE 4096
/* function definitions */
int bot_connect(char *server, unsigned int port, int *sockfd); /* connect */
int bot_send(int sockfd, char *out, int debug); /* send */
int bot_read(int sockfd, char *recvline, int debug); /* read */
void handle_in(int sockfd, char *recvline, int debug); /* handle lines */

int main(int argc, char **argv)
{
	/* 
	   connect the bot to the server and enter a loop
	 */

	/* server definitions */
	char server[] = "172.82.180.204"; /* danieljon.es */
	//char server[] = "188.40.75.16"; /* 3phasegaming.net */
	int port = 6667; /* IRC port */
	/* sockets, buffers and debug definitions */
	int sockfd, in, debug; /* socket, data size and debug */
	char recvline[MAXLINE + 1], out[MAXLINE + 1]; /* receive and out buffers */
	char *pos;
	int joinedchannel = 0; /* ugly hack, fix one day */
	/*
	   set to 1 to toggle debugging
	   when set to 1 extra output information will be printed
	 */
	debug = 0;
	/* attempt connection */
	printf("attempting server connection\n"); 
	if (!bot_connect(server, port, &sockfd))
	{
		printf("Failed to connect to %s, check the IRC server exists.\n", server);
		exit(1);
	}
	/* connection at this point should be made, follow IRC standards */
	printf("connection was successful to %s %i\n", server, port);
	bot_send(sockfd, "USER BeamBot 8 * : Daniel Jones\r\n", debug);
	bot_send(sockfd, "NICK BeamBot\r\n", debug);
	/*
	   now that we are in our channel, enter a loop for commands/pings
	 */
	while(1)
	{
		recvline[0] = 0;
		in = bot_read(sockfd, recvline, debug);
		if (in > 0) {
			recvline[in] = 0;
			/* handle server pings, still runs if someone types PING, fix one day */
			if (strstr(recvline, "PING") != NULL) {
				out[0] = 0;
				pos = strstr(recvline, " ")+1;
				sprintf(out, "PONG %s\r\n", pos);
				bot_send(sockfd, out, debug);
				if (!joinedchannel)
				{
					joinedchannel = 1;
					bot_send(sockfd, "JOIN #tests\r\n", debug);
				}
			}
			/* message isn't a ping, send off to be handled */
			handle_in(sockfd, recvline, debug);
		}
	}
	exit(0);

}

int bot_connect(char *server, unsigned int port, int *sockfd)
{
	/*
	   this function will handle connecting the bot to our IRC server.
	 */
	struct sockaddr_in servaddr;
	bzero(&servaddr, sizeof(servaddr));
	servaddr.sin_family = AF_INET;
	servaddr.sin_port = htons(port);
	if ((*sockfd = socket(AF_INET, SOCK_STREAM, 0)) <= 0)
	{
		return 0;
	}
	if (inet_pton(AF_INET, server, &servaddr.sin_addr) <= 0)
	{
		return 0;
	}
	if (connect(*sockfd, (struct sockaddr*) &servaddr, sizeof(servaddr)) < 0)
	{
		return 0;
	}
	/* assume we are connected if we reach here */
	return 1;
}

int bot_send(int sockfd, char *out, int debug)
{
	/*
	   this function handles sending out through our socket 
	 */
	if (debug)
		printf("SENDING: %s", out);
	return send(sockfd, out, strlen(out), 0);
}

int bot_read(int sockfd, char *recvline, int debug)
{
	/*
	   this function handles receiving from our socket
	 */
	int in;
	in = read(sockfd, recvline, MAXLINE);
	if (in > 0 && debug)
		printf("RECEIVED: %s", recvline);
	return in;
}

void handle_in(int sockfd, char *recvline, int debug)
{
	/*
	   this function will handle incoming messages that aren't pings/pongs
	   here we will format our strings for logging, send them to be logged and get commands parsed and responded to
	 */
	if (strstr(recvline, "PRIVMSG") != NULL)
	{
		/* we received a message, lets handle it */
		if (debug)
			printf("%s\n", recvline);
		//bot_send(sockfd, "PRIVMSG #tests :message\r\n", debug);
	}
}
